### PR TITLE
Activation token in auth service

### PR DESCRIPTION
**Pull request by mvein**

### CODE REVIEW

File: PictureApp.API.Services/AuthService.cs [hunk 1]
```diff
@@ -33,8 +41,29 @@ public async Task Register(UserForRegisterDto userForRegister)
...
+        }
+
+        public async Task Activate(string token)
+        {
+            if (_activationTokenProvider.IsTokenExpired(token))
+            {
+                throw new ArgumentException("Given token is already expired");
+            }
+
+            var activationToken = await _accountActivationTokenRepository.SingleAsync(x => x.Token == token);
```
[the remark this answers is not in the document]
Comment (mvein) [reply PictureApp.API.Services/AuthService.cs]: Sure. I will use SecurityTokenExpiredException. I think, it fits the best here.

File: PictureApp.API.Services/AuthService.cs [hunk 2]
```diff
@@ -33,8 +41,29 @@ public async Task Register(UserForRegisterDto userForRegister)
...
+            var user = await _userRepository.SingleAsync(x => x.Id == activationToken.UserId);
```
[the remark this answers is not in the document]
Comment (mvein) [reply PictureApp.API.Services/AuthService.cs]: Good point!

File: PictureApp.API.Services/AuthService.cs [hunk 3]
```diff
@@ -33,8 +41,29 @@ public async Task Register(UserForRegisterDto userForRegister)
             var userToCreate = _mapper.Map<User>(userForRegister);
             userToCreate.PasswordHash = passwordHash;
             userToCreate.PasswordSalt = passwordSalt;
-            
-            await _repository.AddAsync(userToCreate);
+            var activationToken = new AccountActivationToken
```
[the remark this answers is not in the document]
Comment (mvein) [reply PictureApp.API.Services/AuthService.cs]: Agree. Code become more readable.

File: PictureApp.API.Tests/Providers/ActivationTokenProviderTests.cs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            // ASSERT
+            actual.Should().BeEquivalentTo(expected);
+        }
+
+        [Test]
+        public void IsTokenExpired_WhenCalledAndTokenIsOlderThan24Hours_TrueExpected()
+        {
+            // ARRANGE
+            var provider = new ActivationTokenProvider();
+            var token = provider.CreateToken();
```
[the remark this answers is not in the document]
Comment (mvein) [reply PictureApp.API.Tests/Providers/ActivationTokenProviderTests.cs]: When the token is created current date is taken. Afterwards date is moved to the future in order to check whether token is still valid (not expired). If it is not clear, I can set system date explicitly before creation of a token.

File: PictureApp.API.Tests/Providers/ActivationTokenProviderTests.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            // ASSERT
+            actual.Should().BeTrue();
+        }
+
+        [Test]
+        public void IsTokenExpired_WhenCalledAndTokenIsNotOlderThan24Hours_FalseExpected()
+        {
+            // ARRANGE
+            var provider = new ActivationTokenProvider();
+            var token = provider.CreateToken();
```
[the remark this answers is not in the document]
Comment (mvein) [reply PictureApp.API.Tests/Providers/ActivationTokenProviderTests.cs]: Answer same like above.

**Comment by anowakowski** [comment]
ok i checked notation  .HasAnnotation("ProductVersion", "2.2.0-preview3-35497"); locally on my machine. Everything work fine!

**Pull request merged**